### ISSUE TITLE
error:'could not compile ink_lang'

### ISSUE THREAD

**Issue by GowthamChowdary** [paritytech/ink#506]
I get this when i try to build/test the examples in the tutorial (flipper, incrementer)

```
-----------------------------------------------------------the error----------------------------------------------------------------
Compiling ink_lang v2.1.0 (https://github.com/paritytech/ink?tag=latest-v2#cca31543)
error[E0658]: function pointers cannot appear in constant functions
   --> C:\Users\HP\.cargo\git\checkouts\ink-1add513eda8f5a89\cca3154\lang\src\dispatcher.rs:186:30
    |
186 |               pub const fn new(dispatchable: $dispatchable_fn<Msg, S>) -> Self {
    |                                ^^^^^^^^^^^^
...
237 | / impl_dispatcher_for! {
238 | |     /// Dispatcher for storage preserving messages.
239 | |     struct Dispatcher(DispatchableFn);
240 | | }
    | |_- in this macro invocation
    |
    = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
    = help: add `#![feature(const_fn_fn_ptr_basics)]` to the crate attributes to enable
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0658]: function pointers cannot appear in constant functions
   --> C:\Users\HP\.cargo\git\checkouts\ink-1add513eda8f5a89\cca3154\lang\src\dispatcher.rs:186:30
    |
186 |               pub const fn new(dispatchable: $dispatchable_fn<Msg, S>) -> Self {
    |                                ^^^^^^^^^^^^
...
242 | / impl_dispatcher_for! {
243 | |     /// Dispatcher for potentially storage mutating messages and constructors.
244 | |     struct DispatcherMut(mut DispatchableFnMut);
245 | | }
    | |_- in this macro invocation
    |
    = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
    = help: add `#![feature(const_fn_fn_ptr_basics)]` to the crate attributes to enable
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0658]: mutable references are not allowed in constant functions
   --> C:\Users\HP\.cargo\git\checkouts\ink-1add513eda8f5a89\cca3154\lang\src\dispatcher.rs:186:30
    |
186 |               pub const fn new(dispatchable: $dispatchable_fn<Msg, S>) -> Self {
    |                                ^^^^^^^^^^^^
...
242 | / impl_dispatcher_for! {
243 | |     /// Dispatcher for potentially storage mutating messages and constructors.
244 | |     struct DispatcherMut(mut DispatchableFnMut);
245 | | }
    | |_- in this macro invocation
    |
    = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information
    = help: add `#![feature(const_mut_refs)]` to the crate attributes to enable
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0658`.
error: could not compile `ink_lang`
```

on rustc 1.46.0, win 10.

**Comment by Robbepop** [comment]
You need to use a recent `rustc` nightly version for ink! 2.1.0.

**Comment by GowthamChowdary** [comment]
i am, i ran 
rustup update nightly
rustup update stable.. they're both up to date.

**Comment by Robbepop** [comment]
Ah sorry, forgot about this weird issue. So actually this is even weirder since with a recent nightly there have been some compiler analysis added that forbids what ink! 2.1.0 is doing here so the actual solution is to use the `rustc` nightly version that is or was recommended in the workshop. Just do something like `nightly-2020-06-01` or so and it should probably work because the change is not too new.

**Comment by GowthamChowdary** [comment]
I tried cargo +nightly-2020-06-01 test , worked fine. but when i try to do cargo +nightly-2020-06-01 contract build, it says 
ERROR: Building with xbuild

Caused by:
    `rust-src` component not found. Run `rustup component add rust-src`.
i did run rustup +nightly-2020-06-01 component add rust-src, and got 
"info: component 'rust-src' is up to date"
it just keeps throwing the same error at me

**Comment by riusricardo** [comment]
Just add it to the nightly version you are using.
```
rustup component add rust-src --toolchain nightly-2020-06-01
```

**Comment by GowthamChowdary** [comment]
i did, still getting the same error. 
ERROR: Building with xbuild

Caused by:
    `rust-src` component not found. Run `rustup component add rust-src`.

**Comment by Valentine-Mario** [comment]
I also ran into a similar problem. And following the given instructions above does not seem to solve it.

**Comment by Robbepop** [comment]
We are very sorry for this! It seems to be way harder to find a matching combination of all the required dependencies to get ink! 2.1 running at this point in time:
- ink! 2.1 (itself)
- `cargo-contract` version `0.6.2`
- Substrate 2.0-rc4
- `cargo-xbuild` version `??`
- `rustc` version before or equal to `2020-06-01`

**Comment by Valentine-Mario** [comment]
I guess I'll have to wait to Ink 3.0 to try it out again

**Comment by Robbepop** [comment]
Closed since fixed by ink! 3.0-rc1.
